nf-test side quest (#516)
nextflow-io/training: docs/side_quests/nf-test.md

Testing allows you to systematically check that every part of your pipeline is working as expected. The benefits to a developer of well written tests are huge:

- **Confidence**: Because the tests cover the entire pipeline, you can be be confident changing something doesn't affect anything else
- **Trust**: When multiple developers work on the pipeline, they know the other developers haven't broken the pipeline and every component.
- **Transparency**: The tests show where a pipeline is failing and make it easier to track down the problem. They also function as a form of documentation, showing how to run a process or workflow.
- **Speed**: Because the tests are automated, they can be run very quickly and repeatedly. You can iterate quickly with less fear of introducing new bugs.

There are lots of different types of tests we can write:

1. **Module-level tests**: For individual processes
2. **Workflow-level tests**: For a single workflow
3. **Pipeline-level tests**: For the pipeline as a whole
4. **Performance tests**: For the speed and efficiency of the pipeline

> 5. **Stress tests**: To identify the limits of the pipeline

Reviewer: ```suggestion
5. **Stress tests**: Assessing the pipeline's performance under extreme conditions to determine its limits
```

### Takeaway

You should be able to 'test' a pipeline by manually running it.

### What's next?

Initialize `nf-test`.

---

## 1.0. Initialize `nf-test`

The `nf-test` package provides an initialization command that sets up a few things in order for us to start developing tests for our project.

```bash
nf-test init
```

This should produce the following output:

```bash
🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Project configured. Configuration is stored in nf-test.config
```

It also creates a `tests` directory containing a configuration file stub.

## 1.1. Generate an nf-test

`nf-test` comes with a set of tools for building nf-test files, saving us the majority of the work. These come under the subcommand `generate`. Let's generate a test for the pipeline:

```bash
nf-test generate pipeline main.nf
```

```console title="Output"
> nf-test generate pipeline main.nf

Load source file '/workspaces/training/side-quests/nf-test/main.nf'
Wrote pipeline test file '/workspaces/training/side-quests/nf-test/tests/main.nf.test

SUCCESS: Generated 1 test files.
```

> This will create a `tests` directory with a `main.nf.test` and a `nextflow.config` file. This is our pipeline level test file. If you run `tree tests/` you should see something like this:

Reviewer: The `tests` directory has already been created by `nf-test init`, no? That's what is said above. I also tried it here and saw this.

```suggestion
This will create a `main.nf.test` file within the `tests` directory. This is our pipeline level test file. If you run `tree tests/` you should see something like this:
```

Author: Yes, you're correct- thanks!

Good test names should:

1. Start with "Should" to make it clear what the expected behavior is
2. Describe the specific functionality or scenario being tested
3. Be clear enough that if the test fails, you know what functionality is broken

As we add more assertions and specific test cases later, we'll use these more descriptive names to make it clear what each test is verifying.

## 1.2. Run the test

Let's run the test to see what happens.

```bash
nf-test test tests/main.nf.test
```

```console title="nf-test pipeline fail"
> nf-test test tests/main.nf.test

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Test Workflow main.nf

  Test [693ba951] 'Should run without failures' FAILED (4.652s)

  Assertion failed:

  assert workflow.success
         |        |
         workflow false

  Nextflow stdout:

  ERROR ~ No such file or directory: /workspaces/training/side-quests/nf-test/.nf-test/tests/693ba951a20fec36a5a9292ed1cc8a9f/greetings.csv

   -- Check '/workspaces/training/side-quests/nf-test/.nf-test/tests/693ba951a20fec36a5a9292ed1cc8a9f/meta/nextflow.log' file for details
  Nextflow stderr:

>   Nextflow 24.10.4 is available - Please consider updating your version to it
> 
> 
> 
> 

Reviewer 2: ```suggestion
```
Delete for clarity and to avoid distractions.

Author: Sorry, yes, didn't mean to include

This is just the surface, we can keep writing assertions to check the details of the pipeline, but for now let's move on to testing the internals of the pipeline.

### Takeaway

You know how to write an nf-test for a pipeline.

### What's next?

Learn how to test a Nextflow process.

---

## 2 Test a Nextflow process

We don't have to write tests for every part of the pipeline, but the more tests we have the more comprehensive we can be about the pipeline and the more confident we can be that it's working as expected. In this section we're going to test both processes in the pipeline as individual units.

### 2.1. Test the `sayHello` process

Let's start with the `sayHello` process.

Let's use the `nf-test generate` command again to generate tests for the process.

```bash
nf-test generate process main.nf
```

```console title="Output"
> nf-test generate process main.nf

Load source file '/workspaces/training/side-quests/nf-test/main.nf'
Wrote process test file '/workspaces/training/side-quests/nf-test/tests/main.sayhello.nf.test
Wrote process test file '/workspaces/training/side-quests/nf-test/tests/main.converttoupper.nf.test

SUCCESS: Generated 2 test files.
```

> Let's focus for now and focus on the `sayhello` process in the `main.sayhello.nf.test` file.

Reviewer: ```suggestion
Let's focus for now on the `sayhello` process in the `main.sayhello.nf.test` file.
```

        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }

}
```

As before, we start with the test details, followed by the `when` and `then` blocks. However, we also have an additional `process` block which allows us to define the inputs to the process.

Let's run the test to see if it works.

```bash title="nf-test pipeline pass"
nf-test test tests/main.sayhello.nf.test
```

```console title="Process test fails"
> nf-test test tests/main.sayhello.nf.test

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Test Process sayHello

  Test [1eaad118] 'Should run without failures' FAILED (4.876s)

  Assertion failed:

  assert process.success
         |       |
         |       false
         sayHello

  Nextflow stdout:

  Process `sayHello` declares 1 input channel but 0 were specified

   -- Check script '/workspaces/training/side-quests/nf-test/.nf-test-1eaad118145a1fd798cb07e7dd75d087.nf' at line: 38 or see '/workspaces/training/side-quests/nf-test/.nf-test/tests/1eaad118145a1fd798cb07e7dd75d087/meta/nextflow.log' file for more details
  Nextflow stderr:

>   Nextflow 24.10.4 is available - Please consider updating your version to it
> 
> 
> 
> 

Reviewer 2: ```suggestion
```
Delete for clarity and to avoid distractions.

FAILURE: Executed 1 tests in 4.884s (1 failed)
```

> The test fails because the `sayHello` process declares 1 input channel but 0 were specified. Let's fix that by adding an input to the process. Remember from part 1, our `sayHello` process takes a single value input. We should also fix the test name to better reflect what we're testing.

Reviewer: ```suggestion
The test fails because the `sayHello` process declares 1 input channel but 0 were specified. Let's fix that by adding an input to the process. Remember from Part 1, our `sayHello` process takes a single value input. We should also fix the test name to better reflect what we're testing.
```

We've been capitalizing `Part` in the Hello Nextflow training, so suggesting this here for consistency.

Author: Thanks for highlighting. Actually, referencing part 1 isn't very helpful here, since we're on a side-quest. I probably need to re-introduce the workflow content quickly early on

https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Test Process convertToUpper

  Test [c59b6044] 'Should run without failures and produce correct output' PASSED (1.755s)
  Snapshots:
    1 created [Should run without failures and produce correct output]

Snapshot Summary:
  1 created

SUCCESS: Executed 1 tests in 1.764s
```

Note, we have created a snapshot file for the `convertToUpper` process at `tests/main.converttoupper.nf.test.snap`. If we run the test again, we should see the nf-test passes again.

```bash title="nf-test process convertToUpper pass"
nf-test test tests/main.converttoupper.nf.test
```

```console title="nf-test process convertToUpper pass"
> nf-test test tests/main.converttoupper.nf.test

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Test Process convertToUpper

  Test [c59b6044] 'Should run without failures and produce correct output' PASSED (1.798s)

SUCCESS: Executed 1 tests in 1.811s
```

### Takeaway

You know how to write tests for a Nextflow process and run them.

### What's next?

Learn how to run tests for everything at once!

## 3. Run tests for the entire repository

> Running nf-test on each component is fine, but laborious and error prone. Can't we just test a test on everything at once?

Reviewer: ```suggestion
Running nf-test on each component is fine, but laborious and error prone. Can't we just test everything at once?
```
What do you think?

Let's run the workflow to make sure it's working as expected.

```bash
nextflow run main.nf
```

CONGRATULATIONS! You just ran a unit test!

"Wait, what? I just ran the workflow and it worked! How is that a unit test?"

Good question!

Let's break down what just happened.

You ran the workflow with the default parameters, you confirmed it worked and you're happy with the results. This is the essence of unit testing. As you go through this course you will notice we always run the workflow at the start to confirm everything is set up correctly. This gives us the confidence to continue with the training, knowing the code started in a working state.

Unit testing software essentially does this process for us. Let's replace our simple `nextflow run main.nf` with a standardised test provided by nf-test.

### Takeaway

You should be able to 'test' a pipeline by manually running it.

### What's next?

Initialize `nf-test`.

---

## 1.0. Initialize `nf-test`

The `nf-test` package provides an initialization command that sets up a few things in order for us to start developing tests for our project.

```bash
nf-test init
```

This should produce the following output:

```bash

> 🚀 nf-test 0.9.1

Reviewer 3: can we use the same 0.9.2 version all through the document?

https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Project configured. Configuration is stored in nf-test.config
```

It also creates a `tests` directory containing a configuration file stub.

## 1.1. Generate an nf-test

`nf-test` comes with a set of tools for building nf-test files, saving us the majority of the work. These come under the subcommand `generate`. Let's generate a test for the pipeline:

```bash
nf-test generate pipeline main.nf
```

```console title="Output"
> nf-test generate pipeline main.nf

Load source file '/workspaces/training/side-quests/nf-test/main.nf'
Wrote pipeline test file '/workspaces/training/side-quests/nf-test/tests/main.nf.test

SUCCESS: Generated 1 test files.
```

> This will create a `tests` directory with a `main.nf.test` and a `nextflow.config` file. This is our pipeline level test file. If you run `tree tests/` you should see something like this:

Reviewer 3: does nf-test create the tests/ folder?
I think I noticed a different behavior: https://github.com/askimed/nf-test/issues/286

Author: It does

```console title="Test directory contents"
tests/
├── main.nf.test
└── nextflow.config
```

The `main.nf.test` file is our pipeline level test file. Let's open it up and take a look at the contents.

```groovy title="tests/main.nf.test"
nextflow_pipeline {

    name "Test Workflow main.nf"

>     script "main.nf"

Reviewer 3: that should not work if main.nf is one level up

Author: Pretty sure it did though. I'll double check

Reviewer 3: For me this and my previous comment in line 122 are linked.

Author: confirmed, works just fine

>     test("Should run without failures") {

Reviewer 3: Can we please give a better name for the test?

Author: But then we'd have to tell the users to change it, which I don't think adds anything to the message at this point

Reviewer 3: we should tell users to change it.

Author: What would you like them to change it to?

Reviewer 4: I agree naming the test is an important talking point, and *definitely* introduce having multiple tests on the same thing.

Would also be good to introduce different types of tests (not just identity, but also content).

3. **Pipeline-level tests**: For the pipeline as a whole
4. **Integration tests**: For the interaction of the pipeline with other systems
5. **Performance tests**: For the speed and efficiency of the pipeline
6. **Stress tests**: To identify the limits of the pipeline

[**nf-test**](https://www.nf-test.com/) is a tool that allows you to write module, workflow and pipeline level test. In short, it allows you to systematically check every individual part of the pipeline is working as expected, _in isolation_.

In this part of the training, we're going to show you how to use nf-test to write module-level tests for the three processes in our pipeline.

---

## 0. Warmup

Let's move into the project directory.

```bash
cd side-quests/nf-test
```

The `nf-test` directory has the file content like:

```console title="Directory contents"
nf-test
├── greetings.csv
├── main.nf
└── nextflow.config
```

For a detailed description of the files, see the [warmup from Hello Nextflow](../hello_nextflow/00_orientation.md).

### 0.1. Run the workflow

Let's run the workflow to make sure it's working as expected.

```bash
nextflow run main.nf
```

CONGRATULATIONS! You just ran a unit test!

> "Wait, what? I just ran the workflow and it worked! How is that a unit test?"

Reviewer 4: This is confusing for a couple of reasons. 

1. 'Unit test' is the only type of test not explicitly listed above, and non-programmers will typically have never heard of unit tests.
2. Running an entire workflow is not like a unit test at all; arguably the closest thing to a unit test here is a process-level test.

FWIW when I taught the original nf-test material I made the approximation that process level tests are like unit tests, and that workflow level tests are like integration tests (because they test the integration of all the processes together). I'd like to use that formally in these materials.

Reviewer 4: Semantics is everything when we're deciding what to call things 😆 

I'm going off of what the GATK devs called integration tests, but recognize different groups use different terminologies.

So ok, let's just call this a workflow-level test or something to that effect.

Author: I'm with @vdauwera here actually- testing processes feels like unit testing, workflows are integration tests. 

But maybe avoiding the language altogether and just being specific on what we mean is the best way forward in any case.

Author: Just 'test' is fine here, changed accordingly

Good question!

Let's break down what just happened.

> You ran the workflow with the default parameters, you confirmed it worked and you're happy with the results. This is the essence of unit testing. As you go through this course you will notice we always run the workflow at the start to confirm everything is set up correctly. This gives us the confidence to continue with the training, knowing the code started in a working state.

Reviewer 4: Maybe something like "If you worked through the Hello Nextflow training course, you'll have noticed we always started every section by running the workflow we were using as a starting point, to confirm everything is set up correctly."

```

```console title="nf-test process convertToUpper pass"
> nf-test test tests/main.converttoupper.nf.test

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Test Process convertToUpper

  Test [c59b6044] 'Should run without failures' PASSED (1.755s)
  Snapshots:
    1 created [Should run without failures]

Snapshot Summary:
  1 created

SUCCESS: Executed 1 tests in 1.764s
```

Note, we have created a snapshot file for the `convertToUpper` process at `tests/main.converttoupper.nf.test.snap`. If we run the test again, we should see the nf-test passes again.

```bash title="nf-test process convertToUpper pass"
nf-test test tests/main.converttoupper.nf.test
```

```console title="nf-test process convertToUpper pass"
> nf-test test tests/main.converttoupper.nf.test

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Test Process convertToUpper

  Test [c59b6044] 'Should run without failures' PASSED (1.798s)

SUCCESS: Executed 1 tests in 1.811s
```

### Takeaway

You know how to write tests for a Nextflow process and run them.

### What's next?

Learn how to run tests for everything at once!

> ## 3. Run tests for the entire repo

Reviewer 4: Please write repository in full rather than just repo, as that makes it more understandable for non-English speakers
